### PR TITLE
Use multi-stage build for Docker images

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Following from the discussion on #3602, changing the DockerFile to use multi-stage build. 

This avoids having the Pulsar tgz to be included twice in the image (because of the `mv /apache-pulsar-2.x.y /pulsar`). 

The image size goes from 1G to < 400MB.

### CODE REVIEW

**Comment by aahmed-se** [comment]
run java8 tests

**Comment by aahmed-se** [comment]
run integration tests

**Comment by aahmed-se** [comment]
something is missing
```
2019-05-01\T\00:45:04.560 [ERROR] Failed to execute goal com.spotify:dockerfile-maven-plugin:1.4.9:build (default) on project pulsar-docker-image: Could not build image: The command '/bin/sh -c /pulsar/bin/install-pulsar-client-27.sh' returned a non-zero code: 1 -> [Help 1]
2019-05-01\T\00:45:04.565 [ERROR] 
2019-05-01\T\00:45:04.565 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
2019-05-01\T\00:45:04.565 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
```

**Comment by aahmed-se** [comment]
```
2019-05-01\T\00:44:33.170 [INFO] + pip install /pulsar/pulsar-client/pulsar_client-2.4.0-cp27-cp27mu-manylinux1_x86_64.whl
2019-05-01\T\00:44:33.170 [INFO] 
2019-05-01\T\00:44:33.804 [INFO] ERROR: pulsar_client-2.4.0-cp27-cp27mu-manylinux1_x86_64.whl is not a supported wheel on this platform.
```

File: docker/pulsar/Dockerfile [hunk 1]
```diff
@@ -36,6 +34,25 @@ COPY scripts/set_python_version.sh /pulsar/bin
...
+RUN apt-get update \
+     && apt-get install -y netcat dnsutils \
+                 python2.7 python-setuptools \
+                 python3 python3-setuptools \
+     && apt-get clean \
+     && rm -rf /var/lib/apt/lists/*
+
+RUN easy_install pip
+RUN pip install kazoo pyyaml
+RUN easy_install3 pip
```
Comment (aahmed-se) [inline docker/pulsar/Dockerfile]: need pip3 install kazoo pyyaml

Comment (merlimat) [reply docker/pulsar/Dockerfile]: Do we need both for py2 and py3?

Comment (aahmed-se) [reply docker/pulsar/Dockerfile]: I am in favor of dropping python2. I think everything works with py3 except the dashboard.

**Pull request merged**